pkg/s3select/sql/value.go

```diff
@@ -65,198 +44,214 @@ const (
+}
+
+// SameTypeAs return whether the two types are strictly the same.
+func (v Value) SameTypeAs(b Value) (ok bool) {
+	switch v.value.(type) {
+	case bool:
+		_, ok = b.value.(bool)
+	case string:
+		_, ok = b.value.(string)
+	case int64:
+		_, ok = b.value.(int64)
+	case float64:
+		_, ok = b.value.(float64)
+	case time.Time:
+		_, ok = b.value.(time.Time)
+	case []byte:
+		_, ok = b.value.([]byte)
 	default:
+		ok = reflect.TypeOf(v.value) == reflect.TypeOf(b.value)
 	}
-	return
+	return ok
 }
 
 // ToBool returns the bool value; second return value refers to if the bool
 // conversion succeeded.
-func (v *Value) ToBool() (val bool, ok bool) {
-	switch v.vType {
-	case typeBool:
-		return v.value.(bool), true
-	}
-	return false, false
+func (v Value) ToBool() (val bool, ok bool) {
+	val, ok = v.value.(bool)
+	return
 }
 
 // ToTimestamp returns the timestamp value if present.
-func (v *Value) ToTimestamp() (t time.Time, ok bool) {
-	switch v.vType {
-	case typeTimestamp:
-		return v.value.(time.Time), true
-	}
-	return t, false
+func (v Value) ToTimestamp() (t time.Time, ok bool) {
+	t, ok = v.value.(time.Time)
+	return
 }
 
 // ToBytes converts Value to byte-slice.
-func (v *Value) ToBytes() ([]byte, bool) {
-	switch v.vType {
-	case typeBytes:
-		return v.value.([]byte), true
-	}
-	return nil, false
+func (v Value) ToBytes() (val []byte, ok bool) {
+	val, ok = v.value.([]byte)
+	return
 }
 
 // IsNull - checks if value is missing.
-func (v *Value) IsNull() bool {
-	return v.vType == typeNull
+func (v Value) IsNull() bool {
+	switch v.value.(type) {
+	case nil:
+		return true
+	}
+	return false
 }
 
-func (v *Value) isNumeric() bool {
-	return v.vType == typeInt || v.vType == typeFloat
+func (v Value) isNumeric() bool {
+	switch v.value.(type) {
+	case int64, float64:
+		return true
+	}
+	return false
 }
 
 // setters used internally to mutate values
 
 func (v *Value) setInt(i int64) {
-	v.vType = typeInt
 	v.value = i
 }
 
 func (v *Value) setFloat(f float64) {
-	v.vType = typeFloat
 	v.value = f
 }
 
 func (v *Value) setString(s string) {
-	v.vType = typeString
 	v.value = s
 }
 
 func (v *Value) setBool(b bool) {
-	v.vType = typeBool
 	v.value = b
 }
 
 func (v *Value) setTimestamp(t time.Time) {
-	v.vType = typeTimestamp
 	v.value = t
 }
 
+func (v Value) String() string {
+	return fmt.Sprintf("%#v", v.value)
+}
+
 // CSVString - convert to string for CSV serialization
-func (v *Value) CSVString() string {
-	switch v.vType {
-	case typeNull:
+func (v Value) CSVString() string {
+	switch x := v.value.(type) {
+	case nil:
 		return ""
-	case typeBool:
-		return fmt.Sprintf("%v", v.value.(bool))
-	case typeString:
-		return v.value.(string)
-	case typeInt:
-		return fmt.Sprintf("%v", v.value.(int64))
-	case typeFloat:
-		return fmt.Sprintf("%v", v.value.(float64))
-	case typeTimestamp:
-		return FormatSQLTimestamp(v.value.(time.Time))
-	case typeBytes:
-		return fmt.Sprintf("%v", string(v.value.([]byte)))
+	case bool:
```

Comment:
Do we want to use Sprint(a ...interface{}) here (when we already know the type)?

I guess this may have a (small?) performance penalty?

Reply:
So tempting to do the "Go ternary" here: `return map[bool]string{true: "true", false: "false"}[x]`
`

Reply:
Go hack the language !